File: TsinghuaNet.UWP.Helper/SettingsHelper.vb
Imports Windows.Storage

Public Enum UserTheme
    [Default]
    Light
    Dark
    Auto
End Enum

Public Enum UserContentType
    Line
    Ring
    Water
    Graph
End Enum

Public Module SettingsHelper
    Private values As IPropertySet

    Private Function GetValue(Of T)(key As String, Optional def As T = Nothing) As T
        If values.ContainsKey(key) Then
            Return values(key)
        Else
            Return def
        End If
    End Function

    Private Sub SetValue(Of T)(key As String, value As T)
        If values.ContainsKey(key) Then
            values(key) = value
        Else
            values.Add(key, value)
        End If
    End Sub

    Private Const StoredUsernameKey As String = "Username"
    Private Const AutoLoginKey As String = "AutoLogin"
    Private Const BackgroundAutoLoginKey As String = "BackgroundAutoLogin"
    Private Const BackgroundLiveTileKey As String = "BackgroundLiveTile"
    Private Const ThemeKey As String = "Theme"
    Private Const ContentTypeKey As String = "UserContentType"
    Private Const FluxLimitKey As String = "FluxLimit"

    Sub New()
        values = ApplicationData.Current.LocalSettings.Values
        StoredUsername = GetValue(Of String)(StoredUsernameKey)
        AutoLogin = GetValue(AutoLoginKey, True)
        BackgroundAutoLogin = GetValue(BackgroundAutoLoginKey, True)
        BackgroundLiveTile = GetValue(BackgroundLiveTileKey, True)
        Theme = GetValue(Of Integer)(ThemeKey, UserTheme.Default)
        ContentType = GetValue(Of Integer)(ContentTypeKey, UserContentType.Ring)
        FluxLimit = GetValue(Of Long?)(FluxLimitKey, Nothing)
    End Sub

    Public Sub SaveSettings()
        SetValue(StoredUsernameKey, StoredUsername)
        SetValue(AutoLoginKey, AutoLogin)
        SetValue(BackgroundAutoLoginKey, BackgroundAutoLogin)
        SetValue(BackgroundLiveTileKey, BackgroundLiveTile)
        SetValue(Of Integer)(ThemeKey, Theme)
        SetValue(Of Integer)(ContentTypeKey, ContentType)
        SetValue(FluxLimitKey, FluxLimit)
    End Sub

    Public Property StoredUsername As String

    Public Property AutoLogin As Boolean

    Public Property BackgroundAutoLogin As Boolean

    Public Property BackgroundLiveTile As Boolean

    Public Property Theme As UserTheme

    Public Property ContentType As UserContentType

    Public Property FluxLimit As Long?
End Module

File: TsinghuaNet.Uno/TsinghuaNet.Uno.Helpers.Shared/SettingsHelper.cs
using Windows.Foundation.Collections;
using Windows.Storage;

namespace TsinghuaNet.Uno.Helpers
{
    static class SettingsHelper
    {
        private static IPropertySet values;

        private static T GetValue<T>(string key, T def = default)
        {
            if (values.ContainsKey(key))
                return (T)values[key];
            else
                return def;
        }

        private static void SetValue<T>(string key, T value)
        {
            if (values.ContainsKey(key))
                values[key] = value;
            else
                values.Add(key, value);
        }

        private const string StoredUsernameKey = "Username";
        private const string AutoLoginKey = "AutoLogin";
#if WINDOWS_UWP
        private const string BackgroundAutoLoginKey = "BackgroundAutoLogin";
        private const string BackgroundLiveTileKey = "BackgroundLiveTile";
#endif
        private const string EnableFluxLimitKey = "EnableFluxLimit";
        private const string FluxLimitKey = "FluxLimit";

        static SettingsHelper()
        {
            values = ApplicationData.Current.LocalSettings.Values;
            StoredUsername = GetValue(StoredUsernameKey, string.Empty);
            AutoLogin = GetValue(AutoLoginKey, true);
#if WINDOWS_UWP
            BackgroundAutoLogin = GetValue(BackgroundAutoLoginKey, true);
            BackgroundLiveTile = GetValue(BackgroundLiveTileKey, true);
#endif
            EnableFluxLimit = GetValue<bool>(EnableFluxLimitKey);
            var limit = GetValue<double>(FluxLimitKey);
            FluxLimit = ByteSize.FromGigaBytes(limit);
        }

        public static void SaveSettings()
        {
            SetValue(StoredUsernameKey, StoredUsername);
            SetValue(AutoLoginKey, AutoLogin);
#if WINDOWS_UWP
            SetValue(BackgroundAutoLoginKey, BackgroundAutoLogin);
            SetValue(BackgroundLiveTileKey, BackgroundLiveTile);
#endif
            SetValue(EnableFluxLimitKey, EnableFluxLimit);
            SetValue(FluxLimitKey, FluxLimit.GigaBytes);
        }

        public static string StoredUsername { get; set; }

        public static bool AutoLogin { get; set; }


#if WINDOWS_UWP
        public static bool BackgroundAutoLogin { get; set; }

        public static bool BackgroundLiveTile { get; set; }
#endif

        public static bool EnableFluxLimit { get; set; }

        public static ByteSize FluxLimit { get; set; }
    }
}
